Update monitoring-es-mb.json to correct datatype

ash-darin opened the pull request:
This fixes #116766

Datatype for bytes should not be keyword but (unsigned_)long. Chose unsinged_long as there should be no negative amounts of bytes.

The total can be reported as "max" by systemd in `/sys/fs/cgroup/system.slice/elasticsearch.service/memory.max`  . metricbeat will then return "max". This is the reason, why 

`elasticsearch.node.stats.os.cgroup.memory.limit.bytes`

is of type `keyword` . This is not necessary for `elasticsearch.node.stats.os.cgroup.memory.usage.bytes` , as this should be a dynamic value and never a string.

ash-darin committed:
Update monitoring-es-mb.json

correct datatype

ash-darin committed:
Create 126842.yaml

Add documentation file as per build enviromnent request

ash-darin commented:
@consulthys Of course this breaks the dataview by creating a conflict but I see no way around it. This will stay broken till the old mappings rotate out. 
To my knowledge these fields are not used in any dashboards but I can make no exhaustive search. As they are text fields, using them for any graphing or math would have been impossible anyway beyond simple display / reporting functions. The Stack monitoring UI for elasticsearch does not show memory usage based on cgroups. The dashboards of the elasticsearch integration do not show memory based on cgroups.  This mismatch only showed up because I created custom dashboards, trying to make use of these values as these seem to be the only ones reflecting memory use of my cluster. I have to note that all other uses of byte values in the ECS are long. So them being keyword makes no sense to begin with. In comparison:

> beat.stats.cgroup.memory.mem.usage.bytes

is of type long. The only reason these might be keyword is if long would not have been enough to store the number value  but 2^63-1 bytes is about 9 Exabytes by my calculation, so that can't be it. Thinking about it now, making it unsigned is actually not necessary.

consulthys commented:
Thanks for the details @ash-darin 
First of all, I totally agree that those fields being keywords is a bug. I'll try to get to the genesis of it and get back to you.

ash-darin committed:
Merge branch 'elastic:main' into patch-2

ash-darin committed:
Merge branch 'main' into patch-2

ash-darin committed:
Update monitoring-es-mb.json

accommodate the case that this is -1

ash-darin committed:
revert limits to keword

While the limit of the memory cgroup may be reported as "max" by the system and therefore needs a keyword, the usage should always be reported as a number. The limit is also not dynamic, therefore it should not be necessary to be a number as you would not need it to do calculations (e.g. average).

ash-darin commented:
> @consulthys
> 
> > Thanks for the details @ash-darin First of all, I totally agree that those fields being keywords is a bug. I'll try to get to the genesis of it and get back to you.
> 
> If systemd reserves "max/unlimited" memory for a process, the cgroup file
> 
> `/sys/fs/cgroup/system.slice/elasticsearch.service/memory.max`
> 
> reads:
> 
> `max`
> 
> and metricbeat returns the string "max" instead of something like "-1"

In more detail (I updated the original issue):

`elasticsearch.node.status.os.cgroup.mem.limit.bytes` is derived from the memory.max value from systemd and may be the string "max" and is returned as such by metricbeat. See https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html

> memory.max
>   A read-write single value file which exists on non-root cgroups. The default is “max”.

Unless this is somehow converted to a value (for example "-1"), it makes sense to keep `elasticsearch.node.status.os.cgroup.mem.limit.bytes` as a **keyword** . 

The value `elasticsearch.node.status.os.cgroup.mem.usage.bytes` is derived from `memory.current` which is 

> memory.current
>   A read-only single value file which exists on non-root cgroups.
>   The total amount of memory currently being used by the cgroup and its descendants.

This should therefore always be a number.

The code for this is here:

https://github.com/elastic/elasticsearch/blob/ecf7eae2602f8bcba7bf902a6ccdb0f0284bc338/server/src/main/java/org/elasticsearch/monitor/os/OsProbe.java#L597-L598

ash-darin committed:
Update monitoring-es-mb.json

Change long to unsigned long, in order to anticipate 64bit counters.

ash-darin committed:
Merge branch 'main' into patch-2

dakrone commented:
Drive-by comment here, but this also needs to increment `MonitoringTemplateRegistry.STACK_MONITORING_REGISTRY_VERSION` in order for the new templates to be installed when ES is upgraded.

ash-darin committed:
Merge branch 'elastic:main' into patch-2

ash-darin commented:
> Drive-by comment here, but this also needs to increment `MonitoringTemplateRegistry.STACK_MONITORING_REGISTRY_VERSION` in order for the new templates to be installed when ES is upgraded.

@dakrone 
I have no idea where to do that. I found this references here:

https://github.com/ash-darin/elasticsearch/blob/85d375c5adbdafb013ffdc1871bb912056240ed9/x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/MonitoringTemplateRegistry.java#L81

but I'm not certain this is what I'm supposed to change. Looking at other PRs I could deduce it from them either. If you could provide a link to the relevant code and and possibly a reference in what manner this value is to be changed (just a simple increment?) I'll gladly do that.

dakrone commented:
Sure no problem, so on this line:

https://github.com/elastic/elasticsearch/blob/afd3a426eabdfda7d4fd6b0c52d76162e3c9c47e/x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/MonitoringTemplateRegistry.java#L80

You would increment it by one. So it's currently at `… + 22`, you can change it to `… + 23`.

ash-darin committed:
Merge branch 'main' into patch-2

ash-darin committed:
Update MonitoringTemplateRegistry.java

Increment Version number

ash-darin committed:
Merge branch 'main' into patch-2

ash-darin commented on docs/changelog/126842.yaml:
```diff
@@ -0,0 +1,5 @@
+pr: 126842
+summary: Fix 
```
> This needs to be fixed to have a filled in changelog entry

Thank you, I am unaware of all the specific small changes to properly document a PR. This seems to differ greatly across the elastic stack. Is there some documentation about this that I overlooked? I'm doing this by hand, as I don't do this on an adhoc scale and don't have a complete development environment set up.

ash-darin replied:
I elaborated on the commit message in the changelog. Is that as expected?

ash-darin committed:
Merge branch 'main' into patch-2

ash-darin committed:
Merge branch 'main' into patch-2

ash-darin committed:
Merge branch 'main' into patch-2

ash-darin committed:
elaborate on reason for change

dakrone commented:
buildkite test this

dakrone committed:
Merge branch 'main' into patch-2

dakrone commented:
buildkite test this

ash-darin committed:
Merge branch 'main' into patch-2

ash-darin committed:
Merge branch 'main' into patch-2

ash-darin committed:
Merge branch 'main' into patch-2

ash-darin committed:
Merge branch 'main' into patch-2

ash-darin commented:
apparently nothing that is needed
